### PR TITLE
Do not buffer pending traces for CI Visibility

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
# What Does This Do
Refactors `datadog.trace.core.PendingTrace`:
- an abstract parent class `TraceCollector` is created
- logic that is not related to pending traces buffering is extracted into the parent class
- the code using `PendingTrace` is changed to use `TraceCollector`
- a sibling class `StreamingTraceCollector` is created, which writes spans as they finish
- the new streaming collector is used if CI Visibility is enabled and the tracer writes to CI Test Cycle intake

# Motivation

The trace agent and intake made certain assumptions that a trace segment would be sent as an atom.
The main thing this impacts is likely the service layer metrics.
While PendingTrace does not always guarantee that a trace will be written atomically, it makes a best effort to maintain atomicity whenever possible.

The above concerns are not relevant for CI Visibility if tests data is written to CI Test Cycle intake, and not to the APM intake (which for CI VIs is a legacy case that is only used with older DD agents).
At the same time buffering traces may lead to losing tests data in some edge cases (namely for traces with continuations which finish after the last test has completed).
Taking this into account, the logic was reworked to simply stream spans instead of buffering for CI VIsibility.

Jira ticket: [SDTEST-460]

[SDTEST-460]: https://datadoghq.atlassian.net/browse/SDTEST-460?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by juan-fernandez — APPROVED** [review]

**Comment by mcculls** [comment]
Reminder to get approval from the platform team before merging

File: dd-trace-core/src/jmh/java/datadog/trace/common/writer/ddagent/TraceMapperBenchmark.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public void init(Blackhole blackhole) {
             .build();
 
     DDTraceId traceId = DDTraceId.ONE;
-    PendingTrace pendingTrace = tracer.createTrace(traceId);
+    TraceCollector traceCollector = tracer.createTrace(traceId);
```
Comment (dougqh) [inline dd-trace-core/src/jmh/java/datadog/trace/common/writer/ddagent/TraceMapperBenchmark.java:104]: Maybe we should rename createTrace to createTraceCollector.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java [hunk 2]
```diff
@@ -566,7 +566,7 @@ public RequestContext getRequestContext() {
 
   @Override
   public Integer forceSamplingDecision() {
-    PendingTrace trace = this.context.getTrace();
+    TraceCollector trace = this.context.getTrace();
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java:569]: Probably change to getTraceCollector

**Review by dougqh — APPROVED** [review]
I'd prefer createTrace and getTrace renamed to createTraceCollector and getTraceCollector respectively.

Otherwise, this looks good to me.  I'm glad to see that is mostly a straight-forward refactor.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java]: Changed

File: dd-trace-core/src/jmh/java/datadog/trace/common/writer/ddagent/TraceMapperBenchmark.java (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply dd-trace-core/src/jmh/java/datadog/trace/common/writer/ddagent/TraceMapperBenchmark.java]: Renamed
